### ISSUE TITLE
rp2: some mpremote fs commands hang when creating new thread on v1.22.1

### ISSUE THREAD

**Issue by sourceperl** [micropython/micropython#13476]
First of all, thanks a lot for this awesome project !

Since v1.22.1 on RP2040, I have strange behavior with certain mpremote fs commands like touch, cp or rm. They hang when a task is executed on the 2nd core.

### commands to trigger abnormal behavior
```bash
# ok
mpremote touch :myfile.txt
# launch thread on 2nd core
mpremote run main.py
# now this hang until we reset the board
mpremote touch :myfile.txt
```

### main.py
```python
import _thread


def core1_task():
    while True:
        pass


_thread.start_new_thread(core1_task, ())
```

### versions in use for this test
* MicroPython v1.22.1 on 2024-01-05; SparkFun Pro Micro RP2040 with RP2040.
* mpremote 1.22.0 (host is a Raspberry Pi 5 on Debian 12).

Also, if i do the same test on the same board with micropython v1.21, it works great.

**Comment by peterhinch** [comment]
That `main.py` leaves the RP2 in an abnormal state - `mpremote` assumes that the target is at a normal REPL. If an application runs a task on the second core it should ensure that the task has terminated before returning to the REPL.

**Comment by sourceperl** [comment]
Thank you very much for your quick response.

I just updated my code and mpremote still hang. 

### main.py
```python
import _thread
from time import sleep_ms


class ThreadFlag:
    def __init__(self) -> None:
        self._lock = _thread.allocate_lock()

    def set(self):
        if not self._lock.locked():
            self._lock.acquire()

    def unset(self):
        if self._lock.locked():
            self._lock.release()

    def is_set(self):
        return self._lock.locked()


class Task:
    def __init__(self) -> None:
        self.run_flag = ThreadFlag()

    def run(self):
        while self.run_flag.is_set():
            pass


# init a task on core1
core1_task = Task()
_thread.start_new_thread(core1_task.run, ())

# simulate user code on core0
sleep_ms(500)

# program on core0 say goodbye and notify core1 to do the same
core1_task.run_flag.unset()
```

Maybe there is something I misunderstood ? But what is mysterious to me is that the problem simply does not exist with the previous version of micropython.

**Comment by BillSobel** [comment]
I don't think there is a mystery, the 1.2.1 was a patch dealing with the Pico second core, sounds like the test cases missed an issue.
"This is a patch release to fix a race condition and potential deadlock in the rp2 port's mp_thread_begin_atomic_section() function, when the second core is in use."

@dpgeorge FYI.

**Comment by sourceperl** [comment]
Further testing with MicroPython 1.25.0-preview.273.gb2ce9b6fb (2025-02-08) indicates that functionality has been restored.

Probably fix by #16620.

**Comment by dpgeorge** [comment]
> Probably fix by #16620

Yes, the bug here sounds exactly like #16619, and should indeed be fixed by #16620.
